frontends/p4/structInitializers.h

```diff
@@ -19,34 +19,66 @@ limitations under the License.
 
 #include "ir/ir.h"
 #include "frontends/p4/typeChecking/typeChecker.h"
+#include "frontends/common/resolveReferences/resolveReferences.h"
 
 namespace P4 {
 
 /// Converts some list expressions into struct initializers.
 class CreateStructInitializers : public Transform {
     ReferenceMap* refMap;
     TypeMap* typeMap;
+    IR::IndexedVector<IR::StatOrDecl> *toMove;
  public:
     CreateStructInitializers(ReferenceMap* refMap, TypeMap* typeMap):
             refMap(refMap), typeMap(typeMap) {
         setName("CreateStructInitializers");
         CHECK_NULL(refMap); CHECK_NULL(typeMap);
+        toMove = new IR::IndexedVector<IR::StatOrDecl>();
     }
 
-    const IR::Node* postorder(IR::AssignmentStatement* statement) override;
     const IR::Node* postorder(IR::MethodCallExpression* expression) override;
     const IR::Node* postorder(IR::Operation_Relation* expression) override;
     const IR::Node* postorder(IR::Declaration_Variable* statement) override;
     const IR::Node* postorder(IR::ReturnStatement* statement) override;
-};
+    const IR::Node* postorder(IR::AssignmentStatement* statement) override;
+    const IR::Node* postorder(IR::P4Control* control) override;
 
+    static const IR::Expression* defaultValue(const IR::Type* type, const Util::SourceInfo srcInfo);
+};
+/// first repeat splits default initializations into assignments with the default values
```

Comment:
I would think that these should be two different passes: one which creates a temporary for every structexpression that shows up in an inconvenient place, and one which expands the initializers - which you already wrote. 
This should probably not entail much code writing, just reorganizing the existing code.
What do you think?

Reply:
In fact, this approach may make it easy to handle comparisons too - you will no longer have comparisons with default initializers. Creating the temporaries will require the type to be known - which is only possible if the type of the structexpression can be inferred. So I think the whole structure of the pass will be simplified.